src/main/java/org/jenkinsci/plugins/ParameterizedRemoteTrigger/pipeline/Handle.java

```diff
@@ -0,0 +1,407 @@
+import org.jenkinsci.plugins.ParameterizedRemoteTrigger.RemoteBuildConfiguration;
+import org.jenkinsci.plugins.ParameterizedRemoteTrigger.remoteJob.BuildData;
+import org.jenkinsci.plugins.ParameterizedRemoteTrigger.remoteJob.BuildStatus;
+import org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.Whitelisted;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import net.sf.json.JSONException;
+import net.sf.json.JSONObject;
+
+/**
+ * A handle to the triggered remote build. This handle is used in Pipelines
+ * to have direct access to the (correct) remote build instead of relying on
+ * environment variables (like in a Job). This prevents issues e.g. when triggering
+ * remote jobs in a parallel pipeline step.
+ */
+@ParametersAreNullableByDefault
+public class Handle implements Serializable {
+
+    private static final long serialVersionUID = 4418782245518194292L;
+
+    private final RemoteBuildConfiguration remoteBuildConfiguration;
+    private final String queueId;
+
+    //Available once moved from queue to an executor
+    private BuildData buildData;
+    private BuildStatus buildStatus;
+
+    private String jobName;
+    private String jobFullName;
+    private String jobDisplayName;
+    private String jobFullDisplayName;
+    private String jobUrl;
+
+    /**
+     * The current local Item (Job, Pipeline,...) where this plugin is currently used.
+     */
+    private final String currentItem;
+
+    /* 
+     * The latest log entries from the last called method.
+     * Unfortunately the TaskListener.getLogger() from the StepContext does
+     * not write to the pipeline log anymore since the RemoteBuildPipelineStep
+     * already finished.
+     * TODO: Once we found a way to log to the pipeline log directly we can switch
+     */
+    private String lastLog;
+
+    public Handle(RemoteBuildConfiguration remoteBuildConfiguration, String queueId, @Nonnull String currentItem)
+    {
+        this.remoteBuildConfiguration = remoteBuildConfiguration;
+        this.queueId = queueId;
+        this.buildData = null;
+        this.buildStatus = null;
+        this.lastLog = "";
+        this.currentItem = currentItem;
+        if(trimToNull(currentItem) == null) throw new IllegalArgumentException("currentItem null");
+    }
+
+    /**
+     * Check if the remote build is still queued (not building yet).
+     *
+     * @return true if still queued, false if already running.
+     * @throws IOException
+     *            if there is an error retrieving the remote build number.
+     * @throws InterruptedException
+     *            if any thread has interrupted the current thread.
+     */
+    @Whitelisted
+    public boolean isQueued() throws IOException, InterruptedException {
+        //Return if we already have the buildData
+        if(buildData != null) return false;
+
+        PrintStreamWrapper log = new PrintStreamWrapper();
+        try {
+            //TODO: This currently blocks
```

Comment:
Where? I do not see any synchronization in your code

Reply:
I think the comment means [RemoteBuildConfiguration.getBuildData](https://github.com/sap-production/parameterized-remote-trigger-plugin/blob/e3e6512ee9facb7118909325790dbd00fd5380e9/src/main/java/org/jenkinsci/plugins/ParameterizedRemoteTrigger/RemoteBuildConfiguration.java#L880). This method loops until the build is running (also not perfect) which means `isQueued` will not return before the build is running instead of immediately returning `true` if the build is still queued.